File: Div3/881/D.cpp
#include <bits/stdc++.h>
using namespace std;

#define rep(i, a, b) for(int i = a; i < (b); ++i)
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()
typedef long long ll;
typedef pair<int, int> pii;
typedef vector<int> vi;



vector<vi> adjlist;
vi leaves;
vi parent;

unsigned long long count(int a) {
    if (leaves[a] != 0)
        return leaves[a];
    int num = 0;
    for(int i : adjlist[a]) {
        if(i != parent[a]) {
            num += count(i);
        }
        leaves[a] = num;
    }
    return num;
}



int main() {
	cin.tie(0)->sync_with_stdio(0);
	cin.exceptions(cin.failbit);

    int cases;
    cin >> cases;
    for(int caseNum=0; caseNum<cases; caseNum++) {
        int N;
        cin >> N;
        adjlist = vector<vi>(N+1);

        for(int i=0; i<N-1; i++) {
            int u,v;
            cin >> u >> v;
            adjlist[u].push_back(v);
            adjlist[v].push_back(u);
        }

        leaves = vi(N+1);
        parent = vi(N+1);

        vi dfs({1});
        set<int> seen;
        while(!dfs.empty()) {
            int curr = dfs.back();
            dfs.pop_back();
            seen.insert(curr);
            bool leaf = true;
            for(int i : adjlist[curr]) {
                if(!seen.count(i)) {
                    dfs.push_back(i);
                    parent[i] = curr;
                    leaf = false;
                }
            }

            if(leaf) {
                leaves[curr] = 1;
                // int p = parent[curr];
                // while(p != 0) {
                //     leaves[p]++;
                //     p = parent[p];
                // }
            }
        }

        int Q;
        cin >> Q;
        for (int i=0; i<Q; i++) {
            int a,b;
            cin >> a >> b;
            cout << count(a) * count(b) << endl;
        }
    }
}

File: Div3/881/D.py
from collections import defaultdict

if __name__ == '__main__':
    cases = int(input())
    for _ in range(cases):
        N = int(input())
        adjlist = defaultdict(list)
        for _ in range(N-1):
            u, v = map(int, input().split())
            adjlist[u].append(v)
            adjlist[v].append(u)
        
        leaves = [0 for x in range(N+1)]
        parent = [0 for x in range(N+1)]
        dfs = [1]
        seen = set()
        while dfs:
            curr = dfs.pop()
            seen.add(curr)
            leaf = True
            for i in adjlist[curr]:
                if i not in seen:
                    dfs.append(i)
                    parent[i] = curr
                    leaf = False
            
            if leaf:
                leaves[curr] = 1
            #     numleaves[curr] = 1
            #     p = parent[curr]
            #     while p != 0:
            #         numleaves[p] += 1
            #         p = parent[p]
        # print(parent)
        # print(numleaves)
        
        def count(a):
            if leaves[a] != 0:
                return leaves[a]
            num = 0
            for i in adjlist[a]:
                if i != parent[a]:
                    num += count(i)
            leaves[a] = num
            return num
        Q = int(input())
        for _ in range(Q):
            a, b = map(int, input().split())
            print(count(a) * count(b))